File: Sources/ComicCore/MOBIBook.swift
import Foundation
import ImageIO
import SwiftSoup

/// Unencrypted MOBI 6 image comics, including the MOBI 6 rendition of hybrid books.
/// Record ordinals are never treated as reading order: HTML recindex references are authoritative.
public final class MOBIBook {
    private let data: Data
    private let offsets: [Int]
    public private(set) var units: [ReadingUnit] = []
    public private(set) var title = ""
    public private(set) var warnings: [String] = []
    public private(set) var direction: ReadingDirection?
    private static let textLimit = 32 * 1024 * 1024
    private static func invalid(_ detail: String) -> ReaderError { .message("MOBI 文件无效：\(detail)") }
    private static func number(_ bytes: Data, _ offset: Int, _ width: Int) throws -> Int {
        guard offset >= 0, width > 0, offset <= bytes.count-width else { throw invalid("记录被截断。") }
        return bytes[offset..<offset+width].reduce(0) { ($0 << 8) | Int($1) }
    }
    public init(url: URL) throws {
        data = try Data(contentsOf: url, options: .mappedIfSafe)
        guard data.count >= 78, data[60..<68] == Data("BOOKMOBI".utf8) else { throw Self.invalid("不是 BOOKMOBI 容器。") }
        let count = try Self.number(data,76,2)
        guard count > 1, 78+count*8 <= data.count else { throw Self.invalid("记录表不完整。") }
        var positions: [Int] = []
        for i in 0..<count {
            let offset = try Self.number(data,78+i*8,4)
            guard offset >= 78+count*8, (offset < data.count), offset > (positions.last ?? 0) else { throw Self.invalid("记录偏移越界或无序。") }
            positions.append(offset)
        }
        offsets = positions + [data.count]
        let header = try record(0)
        let compression = try Self.number(header,0,2), textLength = try Self.number(header,4,4)
        let textRecords = try Self.number(header,8,2)
        guard try Self.number(header,12,2) == 0 else { throw ReaderError.message("此 MOBI 受 DRM 保护，无法读取。请使用无 DRM 的漫画文件。") }
        guard header.count >= 116, header[16..<20] == Data("MOBI".utf8) else { throw Self.invalid("缺少 MOBI 头。") }
        let length = try Self.number(header,20,4), version = try Self.number(header,36,4)
        guard length >= 100, length <= header.count-16 else { throw Self.invalid("MOBI 头长度越界。") }
        guard version <= 6 else { throw ReaderError.message("当前支持 MOBI 6 及含 MOBI 6 的双格式漫画；纯 KF8/AZW3 请先转换为 EPUB。") }
        guard compression == 1 || compression == 2 else { throw ReaderError.message("此 MOBI 使用尚未支持的 HUFF/CDIC 或其他压缩方式，请先转换为 EPUB。") }
        guard textLength > 0, textLength <= Self.textLimit, textRecords > 0, textRecords < count else { throw Self.invalid("正文长度或记录数量异常。") }
        let imageStart = try Self.number(header,108,4)
        guard imageStart > textRecords, imageStart < count else { throw Self.invalid("图片记录起点无效。") }
        let encodingID = try Self.number(header,28,4)
        let encoding: String.Encoding
        switch encodingID { case 65001: encoding = .utf8; case 1252: encoding = .windowsCP1252
        default: throw ReaderError.message("暂不支持此 MOBI 的文字编码（\(encodingID)）。") }
        var metadata: [Int: Data] = [:]
        if length >= 116, try Self.number(header,128,4) & 0x40 != 0 {
            let start = 16+length
            guard start+12 <= header.count, header[start..<start+4] == Data("EXTH".utf8) else { throw Self.invalid("EXTH 元数据缺失。") }
            let size = try Self.number(header,start+4,4), entries = try Self.number(header,start+8,4)
            guard size >= 12, size <= header.count-start, entries <= (size-12)/8 else { throw Self.invalid("EXTH 长度无效。") }
            var cursor = start+12
            for _ in 0..<entries {
                guard cursor+8 <= start+size else { throw Self.invalid("EXTH 记录越界。") }
                let type = try Self.number(header,cursor,4), bytes = try Self.number(header,cursor+4,4)
                guard bytes >= 8, bytes <= start+size-cursor else { throw Self.invalid("EXTH 记录长度无效。") }
                metadata[type] = header.subdata(in: cursor+8..<cursor+bytes); cursor += bytes
            }
        }
        let nameOffset = try Self.number(header,84,4), nameLength = try Self.number(header,88,4)
        if nameOffset <= header.count, nameLength <= header.count-nameOffset {
            title = String(data: header.subdata(in: nameOffset..<nameOffset+nameLength), encoding: encoding) ?? ""
        }
        if let value = metadata[503], let name = String(data:value,encoding:encoding), !name.isEmpty { title = name }
        // EXTH 525 is text writing mode, not authoritative physical page progression.
        // Leave navigation direction to the reader preference rather than infer it from writing mode.
        if let boundary = metadata[121], boundary.count == 4, try Self.number(boundary,0,4) != 0xffffffff {
            warnings.append("双格式 MOBI：当前读取兼容正文，KF8 专属布局和局部放大信息未应用。")
        }
        let extra = length >= 228 ? try Self.number(header,240,4) : 0
        var text = Data()
        for i in 1...textRecords {
            try Task.checkCancellation()
            let bytes = try Self.stripTrailer(record(i), flags:extra)
            let decoded = compression == 2 ? try Self.decompress(bytes) : bytes
            guard decoded.count <= Self.textLimit-text.count else { throw Self.invalid("正文解压超过安全上限。") }
            text.append(decoded)
        }
        guard text.count >= textLength else { throw Self.invalid("正文解压长度不足。") }
        guard let html = String(data:text.prefix(textLength),encoding:encoding) else { throw Self.invalid("正文编码损坏。") }
        let document = try SwiftSoup.parse(html)
        guard let body = document.body(), try body.text().trimmingCharacters(in:.whitespacesAndNewlines).isEmpty else {
            throw ReaderError.message("此 MOBI 含文字或混合排版，当前漫画模式不能完整呈现，请转换为 EPUB 阅读。")
        }
        guard try body.select("svg, script, iframe, object").isEmpty() else { throw ReaderError.message("此 MOBI 含复杂页面，当前漫画模式不支持。") }
        let images = try body.select("img").array()
        guard !images.isEmpty, images.count <= 100_000 else { throw Self.invalid("没有可阅读的图片正文。") }
        var refs: [Int?] = []
        for element in images {
            let value = try element.attr("recindex")
            if let index = Int(value), index > 0, index <= count { refs.append(imageStart+index-1) }
            else { refs.append(nil) }
        }
        var cover: Int?
        if let value = metadata[201], value.count == 4 {
            let relative = try Self.number(value,0,4)
            if relative < count-imageStart { cover = imageStart+relative }
        }
        // MOBI readers expose EXTH cover separately; only insert when absent from body.
        if let cover, !refs.contains(where:{$0 == cover}) { units.append(makeUnit(record:cover,occurrence:-1,isCover:true)) }
        for (index, ref) in refs.enumerated() {
            try Task.checkCancellation()
            units.append(makeUnit(record:ref,occurrence:index,isCover:ref != nil && ref == cover))
        }
        if units.contains(where:{$0.error != nil}) { warnings.append("部分图片缺失或无法解码，已保留原阅读位置。") }
        warnings.append("MOBI 图片模式按正文图片顺序阅读，不还原 HTML/CSS 拼版。")
    }
    private func makeUnit(record index: Int?, occurrence: Int, isCover: Bool) -> ReadingUnit {
        let resource = index.map { "mobi:record:\($0)" } ?? "mobi:missing"
        var width = 0.0, height = 0.0, error: String?
        if let index, let bytes = try? record(index), let source = CGImageSourceCreateWithData(bytes as CFData,nil) {
            (width,height) = ImageTools.dimensions(source,index:0)
            if width <= 0 || height <= 0 { error = "MOBI 图片尺寸无效。" }
        } else { error = "MOBI 图片记录缺失或不支持。" }
        return ReadingUnit(locator:SourceLocator(resource:resource,occurrence:occurrence),title:isCover ? "封面" : "第 \(occurrence+1) 页",width:width,height:height,isCover:isCover,error:error)
    }
    public func imageData(for unit: ReadingUnit) throws -> Data {
        guard unit.locator.resource.hasPrefix("mobi:record:"), let index = Int(unit.locator.resource.dropFirst(12)) else { throw Self.invalid("图片定位无效。") }
        return try record(index)
    }
    private func record(_ index: Int) throws -> Data {
        guard index >= 0, index+1 < offsets.count else { throw Self.invalid("图片引用越界。") }
        return data.subdata(in: offsets[index]..<offsets[index+1])
    }
    private static func stripTrailer(_ input: Data, flags: Int) throws -> Data {
        var end = input.count, bits = flags >> 1
        while bits != 0 {
            if bits & 1 != 0 {
                var size = 0, shift = 0, terminated = false
                for i in (max(0,end-4)..<end).reversed() {
                    let byte = Int(input[i]); size |= (byte & 127) << shift; shift += 7
                    if byte & 128 != 0 { terminated = true; break }
                }
                guard terminated, size > 0, size <= end else { throw invalid("正文尾部索引损坏。") }
                end -= size
            }
            bits >>= 1
        }
        if flags & 1 != 0 {
            guard end > 0 else { throw invalid("多字节尾部被截断。") }
            let size = Int(input[end-1] & 3)+1
            guard size <= end else { throw invalid("多字节尾部越界。") }; end -= size
        }
        return input.subdata(in:0..<end)
    }
    private static func decompress(_ input: Data) throws -> Data {
        var output = [UInt8](), i = 0
        while i < input.count {
            let byte = Int(input[i]); i += 1
            if (1...8).contains(byte) {
                guard byte <= input.count-i else { throw invalid("PalmDOC 字面量被截断。") }
                output.append(contentsOf:input[i..<i+byte]); i += byte
            } else if byte < 128 { output.append(UInt8(byte)) }
            else if byte >= 192 { output.append(32); output.append(UInt8(byte ^ 128)) }
            else {
                guard i < input.count else { throw invalid("PalmDOC 回引用被截断。") }
                let value = (byte << 8) | Int(input[i]); i += 1
                let distance = (value >> 3) & 2047, length = (value & 7)+3
                guard distance > 0, distance <= output.count else { throw invalid("PalmDOC 回引用越界。") }
                for _ in 0..<length { output.append(output[output.count-distance]) }
            }
            guard output.count <= 65536 else { throw invalid("PalmDOC 记录解压超过上限。") }
        }
        return Data(output)
    }
}

File: windows/ShuiMan.Core/MobiBook.cs
using System.Buffers.Binary;
using System.IO;
using System.Text;
using AngleSharp.Html.Parser;

namespace ShuiMan.Core;

/// <summary>Unencrypted MOBI 6 image comics, including the MOBI 6 half of hybrid books.</summary>
internal sealed class MobiBook : RasterBook
{
    private const int TextLimit = 32 * 1024 * 1024;
    private readonly byte[] data;
    private readonly int[] offsets;
    private readonly Dictionary<int, (RasterInfo? Info, string? Error)> infos = [];

    public MobiBook(string path, CancellationToken token)
    {
        Publication = DocumentEngine.Describe(path, "mobi");
        data = DocumentEngine.ReadFile(path, 512L * 1024 * 1024);
        if (data.Length < 78 || !data.AsSpan(60, 8).SequenceEqual("BOOKMOBI"u8)) throw Invalid("不是 BOOKMOBI 容器。");
        int count = Number(data, 76, 2);
        if (count < 2 || 78 + count * 8 > data.Length) throw Invalid("记录表不完整。");
        offsets = new int[count + 1];
        for (int i = 0; i < count; i++)
        {
            int offset = Number(data, 78 + i * 8, 4);
            if (offset < 78 + count * 8 || offset >= data.Length || (i > 0 && offset <= offsets[i - 1])) throw Invalid("记录偏移越界或无序。");
            offsets[i] = offset;
        }
        offsets[count] = data.Length;
        var header = Record(0);
        int compression = Number(header, 0, 2), textLength = Number(header, 4, 4), textRecords = Number(header, 8, 2);
        if (Number(header, 12, 2) != 0) throw new NotSupportedException("此 MOBI 受 DRM 保护，无法读取。请使用无 DRM 的漫画文件。");
        if (header.Length < 116 || !header.AsSpan(16, 4).SequenceEqual("MOBI"u8)) throw Invalid("缺少 MOBI 头。");
        int length = Number(header, 20, 4), version = Number(header, 36, 4);
        if (length < 100 || length > header.Length - 16) throw Invalid("MOBI 头长度越界。");
        if (version > 6) throw new NotSupportedException("当前支持 MOBI 6 及含 MOBI 6 的双格式漫画；纯 KF8/AZW3 请先转换为 EPUB。");
        if (compression is not (1 or 2)) throw new NotSupportedException("此 MOBI 使用尚未支持的 HUFF/CDIC 或其他压缩方式，请先转换为 EPUB。");
        if (textLength <= 0 || textLength > TextLimit || textRecords <= 0 || textRecords >= count) throw Invalid("正文长度或记录数量异常。");
        int imageStart = Number(header, 108, 4);
        if (imageStart <= textRecords || imageStart >= count) throw Invalid("图片记录起点无效。");
        Encoding.RegisterProvider(CodePagesEncodingProvider.Instance);
        int encodingId = Number(header, 28, 4);
        Encoding encoding = encodingId switch
        {
            65001 => new UTF8Encoding(false, true), 1252 => Encoding.GetEncoding(1252),
            _ => throw new NotSupportedException($"暂不支持此 MOBI 的文字编码（{encodingId}）。")
        };
        var metadata = new Dictionary<int, byte[]>();
        if (length >= 116 && (Number(header, 128, 4) & 0x40) != 0)
        {
            int start = 16 + length;
            if (start + 12 > header.Length || !header.AsSpan(start, 4).SequenceEqual("EXTH"u8)) throw Invalid("EXTH 元数据缺失。");
            int size = Number(header, start + 4, 4), entries = Number(header, start + 8, 4);
            if (size < 12 || size > header.Length - start || entries > (size - 12) / 8) throw Invalid("EXTH 长度无效。");
            int cursor = start + 12;
            for (int i = 0; i < entries; i++)
            {
                if (cursor + 8 > start + size) throw Invalid("EXTH 记录越界。");
                int type = Number(header, cursor, 4), bytes = Number(header, cursor + 4, 4);
                if (bytes < 8 || bytes > start + size - cursor) throw Invalid("EXTH 记录长度无效。");
                metadata[type] = header.AsSpan(cursor + 8, bytes - 8).ToArray();
                cursor += bytes;
            }
        }
        int nameOffset = Number(header, 84, 4), nameLength = Number(header, 88, 4);
        if (nameOffset <= header.Length && nameLength <= header.Length - nameOffset)
        {
            string title = encoding.GetString(header, nameOffset, nameLength).Trim('\0', ' ', '\n', '\r');
            if (title.Length > 0) Publication.Title = title;
        }
        if (metadata.TryGetValue(503, out var titleBytes))
        {
            var title = encoding.GetString(titleBytes).Trim();
            if (title.Length > 0) Publication.Title = title;
        }
        // EXTH 525 is writing mode, not authoritative physical page progression.
        if (metadata.TryGetValue(121, out var boundary) && boundary.Length == 4 && BinaryPrimitives.ReadUInt32BigEndian(boundary) != uint.MaxValue)
            Publication.Warnings.Add("双格式 MOBI：当前读取兼容正文，KF8 专属布局和局部放大信息未应用。");
        uint extra = length >= 228 ? (uint)Number(header, 240, 4) : 0;
        using var text = new MemoryStream();
        for (int i = 1; i <= textRecords; i++)
        {
            token.ThrowIfCancellationRequested();
            var record = StripTrailer(Record(i), extra);
            var decoded = compression == 2 ? Decompress(record) : record;
            if (decoded.Length > TextLimit - text.Length) throw Invalid("正文解压超过安全上限。");
            text.Write(decoded);
        }
        if (text.Length < textLength) throw Invalid("正文解压长度不足。");
        var document = new HtmlParser().ParseDocument(encoding.GetString(text.GetBuffer(), 0, textLength));
        var body = document.Body;
        if (body == null || !string.IsNullOrWhiteSpace(body.TextContent))
            throw new NotSupportedException("此 MOBI 含文字或混合排版，当前漫画模式不能完整呈现，请转换为 EPUB 阅读。");
        if (body.QuerySelector("svg, script, iframe, object, canvas, table, picture, video, audio") != null)
            throw new NotSupportedException("此 MOBI 含复杂页面，当前漫画模式不支持。请转换为 EPUB 阅读。");
        var images = body.QuerySelectorAll("img");
        if (images.Length == 0 || images.Length > 100_000) throw Invalid("没有可阅读的图片正文。");
        var refs = images.Select(image => int.TryParse(image.GetAttribute("recindex"), out int index) && index > 0 && index <= count ? (int?)(imageStart + index - 1) : null).ToList();
        int? cover = null;
        if (metadata.TryGetValue(201, out var coverData) && coverData.Length == 4)
        {
            uint relative = BinaryPrimitives.ReadUInt32BigEndian(coverData);
            if (relative < count - imageStart) cover = imageStart + (int)relative;
        }
        if (cover != null && !refs.Contains(cover)) Publication.Units.Add(MakeUnit(cover, -1, true));
        for (int i = 0; i < refs.Count; i++)
        {
            token.ThrowIfCancellationRequested();
            Publication.Units.Add(MakeUnit(refs[i], i, refs[i] != null && refs[i] == cover));
        }
        if (Publication.Units.Any(x => x.Error != null)) Publication.Warnings.Add("部分图片缺失或无法解码，已保留原阅读位置。");
        Publication.Warnings.Add("MOBI 图片模式按正文图片顺序阅读，不还原 HTML/CSS 拼版。");
    }

    private ReadingUnit MakeUnit(int? index, int occurrence, bool cover)
    {
        string resource = index != null ? $"mobi:record:{index}" : "mobi:missing";
        (RasterInfo? Info, string? Error) result;
        if (index == null) result = (null, "MOBI 图片记录缺失。");
        else if (!infos.TryGetValue(index.Value, out result))
        {
            try { result = (RasterDecoder.Describe(Record(index.Value)).First(), null); }
            catch (Exception ex) when (ex is not OperationCanceledException) { result = (null, $"MOBI 图片无法读取：{ex.Message}"); }
            infos[index.Value] = result;
        }
        return new ReadingUnit
        {
            Locator = new SourceLocator(resource, occurrence), Title = cover ? "封面" : $"第 {occurrence + 1} 页",
            Width = result.Info?.Width ?? 0, Height = result.Info?.Height ?? 0, IsCover = cover, Error = result.Error
        };
    }
    private byte[] Record(int index)
    {
        if (index < 0 || index + 1 >= offsets.Length) throw Invalid("图片引用越界。");
        int length = offsets[index + 1] - offsets[index];
        if (length > BookArchive.ResourceLimit) throw Invalid("单个记录超过读取上限。");
        return data.AsSpan(offsets[index], length).ToArray();
    }
    private static byte[] StripTrailer(byte[] input, uint flags)
    {
        int end = input.Length;
        for (uint bits = flags >> 1; bits != 0; bits >>= 1)
        {
            if ((bits & 1) == 0) continue;
            int size = 0, shift = 0;
            bool terminated = false;
            for (int i = end - 1; i >= Math.Max(0, end - 4); i--)
            {
                int value = input[i]; size |= (value & 127) << shift; shift += 7;
                if ((value & 128) != 0) { terminated = true; break; }
            }
            if (!terminated || size <= 0 || size > end) throw Invalid("正文尾部索引损坏。");
            end -= size;
        }
        if ((flags & 1) != 0)
        {
            if (end == 0) throw Invalid("多字节尾部被截断。");
            int size = (input[end - 1] & 3) + 1;
            if (size > end) throw Invalid("多字节尾部越界。");
            end -= size;
        }
        return input.AsSpan(0, end).ToArray();
    }
    private static byte[] Decompress(byte[] input)
    {
        var output = new List<byte>();
        int i = 0;
        while (i < input.Length)
        {
            int value = input[i++];
            if (value is >= 1 and <= 8)
            {
                if (value > input.Length - i) throw Invalid("PalmDOC 字面量被截断。");
                for (int j = 0; j < value; j++) output.Add(input[i++]);
            }
            else if (value < 128) output.Add((byte)value);
            else if (value >= 192) { output.Add(32); output.Add((byte)(value ^ 128)); }
            else
            {
                if (i >= input.Length) throw Invalid("PalmDOC 回引用被截断。");
                int reference = (value << 8) | input[i++], distance = (reference >> 3) & 2047, length = (reference & 7) + 3;
                if (distance == 0 || distance > output.Count) throw Invalid("PalmDOC 回引用越界。");
                for (int j = 0; j < length; j++) output.Add(output[output.Count - distance]);
            }
            if (output.Count > 65536) throw Invalid("PalmDOC 记录解压超过上限。");
        }
        return output.ToArray();
    }
    private static int Number(byte[] bytes, int offset, int width)
    {
        if (offset < 0 || width <= 0 || offset > bytes.Length - width) throw Invalid("记录被截断。");
        uint value = 0;
        for (int i = 0; i < width; i++) value = (value << 8) | bytes[offset + i];
        if (value > int.MaxValue) throw Invalid("记录数值超出范围。");
        return (int)value;
    }
    private static InvalidDataException Invalid(string message) => new($"MOBI 文件无效：{message}");
    public override byte[]? Resource(string path)
    {
        ObjectDisposedException.ThrowIf(Disposed, this);
        return path.StartsWith("mobi:record:", StringComparison.Ordinal) && int.TryParse(path[12..], out int index) ? Record(index) : null;
    }
}
